File: for-local-test/python/206.reverse-linked-list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if(head == None):
            return None
        prev = None
        return self.reverseListHelper(prev, head)
    def reverseListHelper(self, prev: ListNode, curr: ListNode) -> ListNode:
        if(curr == None):
            return prev
        temp = curr.next
        curr.next = prev
        return self.reverseListHelper(curr, temp)
    
if __name__ == "__main__":
    solution = Solution()
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    reversed_head = solution.reverseList(head)
    while(reversed_head != None):
        print(reversed_head.val) # 5 4 3 2 1
        reversed_head = reversed_head.next

File: for-local-test/206.reverse-linked-list.java

class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }

class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) return null;
        ListNode prev = null;
        return reverseListHelper(prev, head);
    }
    public ListNode reverseListHelper(ListNode prev, ListNode curr){
        if(curr == null) return prev;
        ListNode temp = curr.next;
        curr.next = prev;
        return reverseListHelper(curr, temp);
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        ListNode reversedHead = sol.reverseList(head);
        while(reversedHead != null) {
            System.out.print(reversedHead.val + " ");
            reversedHead = reversedHead.next;
        }
    }
}


